components/authz-service/server/v2/authz.go

```diff
@@ -235,3 +234,68 @@ func NewSwitch(c chan api.Version) *VersionSwitch {
 	}()
 	return &x
 }
+
+func isBeta2p1(version api.Version) bool {
+	return version.Major == api.Version_V2 && version.Minor == api.Version_V1
+}
+
+// setRequestProjects sets the project filter based on the current IAM version
+func (s *authzServer) setRequestProjects(ctx context.Context, v api.Version, p []string) ([]string, error) {
+	var projects []string
+	// while on v2.1, either the requested projects or a complete list of projects
+	// is passed on to the engine
+	if isBeta2p1(v) {
+		if len(p) == 0 {
+			// we make this extra call to cover the case when the following are true:
+			// - no project filter has been provided
+			// - one statement allows All Projects for the given resource/action
+			// - at least one statement denies 1 or more projects for the same resource/action
+			// in order to return a complete list of projects exclusive of the denied projects,
+			// we must provide the engine with the list of all projects instead of an empty list
+			list, err := s.projects.ListProjects(ctx, &api.ListProjectsReq{})
+			if err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+			var projectIDs []string
+			for _, project := range list.Projects {
+				projectIDs = append(projectIDs, project.Id)
+			}
+			projects = projectIDs
+		} else {
+			// we pass the requested projects as is
+			projects = p
+		}
+	}
+
+	// while on v2, we always pass an empty project list to the engine because projects
+	// can't have an effect on the permission decision
+	if !isBeta2p1(v) {
+		// we override the requested projects because no filter should be applied on v2
+		projects = []string{}
+	}
+
+	return projects, nil
+}
+
+// parseAuthorizedProjects sets the engine's allowed project response based on the current IAM version
+func parseAuthorizedProjects(v api.Version, p []string) []string {
+	authorizedProjects := p
+	if !isBeta2p1(v) && len(p) > 0 {
+		// as long as at least one project is authorized
+		// we override the filtered projects with All Projects
+		// because no project filter should be applied on v2.0
```

Comment:
I'd change this comment to `because no project filter should be applied unless on v2.1` to be consistent with other logic/naming.

Reply:
took this out